pkg/storagev2/inmemchunkstore/inmemchunkstore.go

```diff
@@ -13,49 +14,119 @@ import (
 )
 
 type ChunkStore struct {
-	chunks sync.Map
+	chunks map[string]chunkData
+	lock   sync.Mutex
+}
+
+type chunkData struct {
+	chunk  swarm.Chunk
+	stamps []swarm.Stamp
 }
 
 func New() *ChunkStore {
-	return &ChunkStore{}
+	return &ChunkStore{
+		chunks: make(map[string]chunkData),
+	}
+}
+
+func (c *ChunkStore) Get(ctx context.Context, addr swarm.Address) (swarm.Chunk, error) {
+	return c.GetWithStamp(ctx, addr, nil)
 }
 
-func (c *ChunkStore) Get(_ context.Context, addr swarm.Address) (swarm.Chunk, error) {
-	val, found := c.chunks.Load(addr.ByteString())
-	if !found {
+func (c *ChunkStore) GetWithStamp(ctx context.Context, addr swarm.Address, batchID []byte) (swarm.Chunk, error) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	data, ok := c.chunks[addr.ByteString()]
+	if !ok {
 		return nil, storage.ErrNotFound
 	}
-	return val.(swarm.Chunk), nil
+
+	// when batchID is not specified, chunk with first stamp is returned
+	if batchID == nil {
+		return makeChunk(data.chunk, data.stamps[0]), nil
+	}
+
+	// when batchID is specified, we need to search stamps by batchID
+	if st, found := findStampWithBatchID(data.stamps, batchID); found {
+		return makeChunk(data.chunk, st), nil
+	}
+
+	return nil, storage.ErrNotFound
 }
 
 func (c *ChunkStore) Put(_ context.Context, ch swarm.Chunk) error {
-	c.chunks.Store(ch.Address().ByteString(), ch)
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	data, ok := c.chunks[ch.Address().ByteString()]
+	if !ok {
+		data = chunkData{
+			chunk:  makeChunk(ch, nil),
+			stamps: make([]swarm.Stamp, 0, 1),
+		}
+	}
+
+	// append new stamp only if it doesn't exist
+	if _, found := findStampWithBatchID(data.stamps, ch.Stamp().BatchID()); !found {
+		data.stamps = append(data.stamps, ch.Stamp())
+		c.chunks[ch.Address().ByteString()] = data
+	}
+
 	return nil
 }
 
 func (c *ChunkStore) Has(_ context.Context, addr swarm.Address) (bool, error) {
-	_, exists := c.chunks.Load(addr.ByteString())
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	_, exists := c.chunks[addr.ByteString()]
+
 	return exists, nil
 }
 
 func (c *ChunkStore) Delete(_ context.Context, addr swarm.Address) error {
-	c.chunks.Delete(addr.ByteString())
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	delete(c.chunks, addr.ByteString())
+
 	return nil
 }
 
 func (c *ChunkStore) Iterate(_ context.Context, fn storage.IterateChunkFn) error {
-	var retErr error
-	c.chunks.Range(func(_, val interface{}) bool {
-		stop, err := fn(val.(swarm.Chunk))
-		if err != nil {
-			retErr = err
-			return false
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	for _, data := range c.chunks {
+		for _, stamp := range data.stamps {
```

Comment:
So this is not required. We should not treat chunks with multiple stamps as separate chunks for iteration. If a particular stamp is desired, the user has to use the GetWithStamp function. The Iterate function will be provided through some component using the GetWithStamp, so stamp-related information is not required here. 

Depending on how things look, we might not even have Iterate as a core API on ChunkStore, but a separate interface implemented by some component that uses the ChunkStore. For this PR, we should simply return the first stamp we have for the chunk.